Snek.cs

```diff
@@ -0,0 +1,181 @@
+    }
+
+    // methods now
+    public void Move(int gridRows, int gridCols)
+    {
+        snakeRow += D[(int)direction, 0];
+        snakeCol += D[(int)direction, 1];
+
+        // wrapping around situ
+        if (snakeRow >= gridRows)
+        {
+            // Console.WriteLine("The snake has gone too far down");
+            snakeRow = 0;
+        }
+        if (snakeRow < 0)
+        {
+            // Console.WriteLine("The snake has gone too far up");
+            snakeRow = gridRows - 1;
+        }
+        if (snakeCol >= gridCols)
+        {
+            // Console.WriteLine("The snake has gone too far right");
+            snakeCol = 0;
+        }
+        if (snakeCol < 0)
+        {
+            // Console.WriteLine("The snake has gone too far left");
+            snakeCol = gridCols - 1;
+        }
+
+        //keeps track of all of the snake
+        int a = prevLocRow.Length - 2;
+        while (a >= 0)
+
+        {
+            prevLocRow[a + 1] = prevLocRow[a];
+            prevLocCol[a + 1] = prevLocCol[a];
+            a = a - 1;
+        }
+
+        prevLocRow[0] = snakeRow;
+        prevLocCol[0] = snakeCol;
+    }
+
+    public void SetDirection(Direction direction)
+    {
+        this.direction = direction;
+    }
+
+    public bool HasEaten(int appleRow, int appleCol)
+    {
+        // This only checks the head of the snake. 
+        if ((snakeRow == appleRow) && (snakeCol == appleCol))
+        {
+            return true;
+        }
+        return false;
+    }
+
+    public bool Overlaps(int row, int col, ref int score)
+    {
+        // 1. Check the head 
+        if (snakeRow == row && snakeCol == col)
+        {
+            score += 1;
+            Console.WriteLine("Score:" + score);
+            return true;
+        }
+        else    // 2. Check the rest of the body through loop with PrevLoc
+        {
+            for (int i = 0; i < score; i++)
+            {
+                if (prevLocRow[i] == row && prevLocCol[i] == col)
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+    }
+
+    //     public bool CheckOverlap(int score)
+    //     {
+    //         int b = 2; // this is set as 2 as otherwise it just ignores the last part of the tail 
+    //         while (b <= score)
+    //         {
+    //             if (snakeRow == prevLocRow[b] && snakeCol == prevLocCol[b])
+    //             {
+    //                 return true;
+    //             }
+    //             b = b + 1;
+    //         }
+    //         return false;
+    //     }
+
+    // }
+    /// //////////////////////////////////////////
+    public class Apple
+    {
+        public int appleRow;
+        public int appleCol;
+
+        static Random rng = new Random();
+
+        // constructor:
+        public Apple(Snek snake, int gridRows, int gridCols, int score)
+        {
+            this.Respawn(snake, gridRows, gridCols, score);
```

Comment:
```suggestion
            self.gridRows = gridRows;
            self.gridCols = gridCols;
            this.Respawn(snake, score);
```
Maybe saving the grid information that the apple spawns on makes sense given that you'll have to pass those things everywhere whenever you use something like Respawn? 

Where else can you save things like this so that your method calls looks simpler?